src/libdredd/src/mutation_replace_unary_operator.cc

```diff
@@ -0,0 +1,223 @@
+                 << mutant_offset << ")) return false;\n";
+    mutant_offset++;
+  }
+
+  new_function << "  return ";
+  if (IsPrefix(unary_operator_.getOpcode())) {
+    new_function
+        << clang::UnaryOperator::getOpcodeStr(unary_operator_.getOpcode()).str()
+        << "arg();\n";
+  } else {
+    new_function
+        << "arg()"
+        << clang::UnaryOperator::getOpcodeStr(unary_operator_.getOpcode()).str()
+        << ";\n";
+  }
+  new_function << "}\n\n";
+
+  // The function captures |mutant_offset| different mutations, so bump up
+  // the mutation id accordingly.
+  mutation_id += mutant_offset;
+
+  return new_function.str();
+}
+
+void MutationReplaceUnaryOperator::Apply(
+    clang::ASTContext& ast_context, const clang::Preprocessor& preprocessor,
+    int first_mutation_id_in_file, int& mutation_id, clang::Rewriter& rewriter,
+    std::unordered_set<std::string>& dredd_declarations) const {
+  std::string new_function_name = "__dredd_replace_unary_operator_";
+
+  // A string corresponding to the unary operator forms part of the name of the
+  // mutation function, to differentiate mutation functions for different
+  // operators
+  switch (unary_operator_.getOpcode()) {
+    case clang::UnaryOperatorKind::UO_Plus:
+      new_function_name += "Plus";
+      break;
+    case clang::UnaryOperatorKind::UO_Minus:
+      new_function_name += "Minus";
+      break;
+    case clang::UnaryOperatorKind::UO_Not:
+      new_function_name += "Not";
+      break;
+    case clang::UnaryOperatorKind::UO_PreDec:
+      new_function_name += "PreDec";
+      break;
+    case clang::UnaryOperatorKind::UO_PostDec:
+      new_function_name += "PostDec";
+      break;
+    case clang::UnaryOperatorKind::UO_PreInc:
+      new_function_name += "PreInc";
+      break;
+    case clang::UnaryOperatorKind::UO_PostInc:
+      new_function_name += "PostInc";
+      break;
+    case clang::UnaryOperatorKind::UO_LNot:
+      new_function_name += "LNot";
+      break;
+    default:
+      assert(false && "Unsupported opcode");
+  }
+
+  std::string result_type = unary_operator_.getType()
+                                ->getAs<clang::BuiltinType>()
+                                ->getName(ast_context.getPrintingPolicy())
+                                .str();
+  std::string input_type = unary_operator_.getSubExpr()
+                               ->getType()
+                               ->getAs<clang::BuiltinType>()
+                               ->getName(ast_context.getPrintingPolicy())
+                               .str();
+
+  // To avoid problems of ambiguous function calls, the argument types (ignoring
+  // whether they are references or not) are baked into the mutation function
+  // name. Some type names have space in them (e.g. 'unsigned int'); such spaces
+  // are replaced with underscores.
+  new_function_name += "_" + SpaceToUnderscore(input_type);
+
+  if (unary_operator_.isArithmeticOp() ||
```

Comment:
I think it's wrong to include "isArithmeticOp" here.

Unary operators are a bit more complex than binary operators, due to the increment operators. My understanding is as follows:

- The pre operators take an l-value and return an l-value, e.g. you can write: "++x = 2"
- The post operators take an l-value and return an r-value, e.g. you cannot write "x++ = 2"

As a result, we need to treat the input type and result type differently.

I think we can do the followin:

- If unary_operator_>getSubExpr()->isLValue(), make the input type a reference and check if it's volatile-qualified, adding volatile if it is

- if unary_operator->isLValue(), (same)

I *think* this can be done without reference to the particular operator being used and just by checking isLValue(). The same could also be done for the binary operator case I guess, but let's do that separately if we decide to.

Reply:
Addressed.